src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,63 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
 
+    public MateStack() {
+        size = 0;
+    }
+
+    public void push(T value) {
+        if (size == 0) {
+            head = new Node<>(value, null);
+            tail = head;
+            size++;
+            return;
+        }
+        tail.next = new Node<>(value, null);
+        tail = tail.next;
+        size++;
     }
 
     public T peek() {
-        return null;
+        checkStack();
+        return tail.value;
     }
 
     public T pop() {
-        return null;
+        checkStack();
+        Node<T> localNode = head;
+        size--;
+        for (int i = 0; i < size - 1; i++) {
+            localNode = localNode.next;
+        }
+        T result = tail.value;
+        tail = localNode;
+        tail.next = null;
+        return result;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void checkStack() {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+
+        public Node(T value, Node<T> next) {
```

Comment:
I don't think that making this constructor `public` is a good idea. I would recomment changing the access modifier to `default`.